### ISSUE TITLE
Limitations of LocalStack on arm64

### ISSUE THREAD

**Issue by thrau** [localstack/localstack#4921]
**WORK IN PROGRESS**

---

This issue is to collect the current limitations of the arm64 support of the multi-platform docker image.

## Major 

* Without amd64 emulation on an arm64 host, Lambdas running in Docker executor will likely fail. When running the LocalStack arm64 image on an M1, the lambda containers will be run using the emulated amd64 images, and should mostly run.
* **TODO**

### Apple M1 (with amd64 emulation of lambda containers)

- [ ] `FAILED tests/integration/test_apigateway.py::TestWebSockets::test_websocket_api[AWS_PROXY-False-True]`
- [ ] `FAILED tests/integration/test_athena.py::TestAthena::test_simple_query - asse...`
- [ ] `FAILED tests/integration/test_batch.py::TestBatch::test_create_with_additional_config`
- [ ] `FAILED tests/integration/test_events.py::test_api_destinations_with_basic_auth`
- [ ] `FAILED tests/integration/test_glue.py::TestGlue::test_import_data_catalog - A...`
- [ ] `FAILED tests/integration/test_kinesis_analytics.py::TestKinesisAnalytics::test_run_query`
- [ ] `FAILED tests/integration/test_kinesis_analytics.py::TestKinesisAnalytics::test_run_siddhi_query`
- [ ] `FAILED tests/integration/test_lambda.py::test_function_using_layer - botocore...`
- [ ] `FAILED tests/integration/test_rds.py::TestRDS::test_create_snapshot - Asserti...`
- [ ] `FAILED tests/integration/test_rds.py::TestRDS::test_query_from_s3 - botocore....`
- [ ] `FAILED tests/integration/test_redshift.py::TestRedshift::test_create_cluster`
- [ ] `FAILED tests/integration/test_transfer.py::TestAWSTransfer::test_basic_transfer_api`
- [ ] `FAILED tests/integration/cloudformation/test_cloudformation_msk.py::test_create_cluster`
- [ ] `ERROR tests/integration/test_serverless.py::TestServerless::test_websocket_deployed`

### Ampere Altra (without emulation)

- [ ] `FAILED tests/integration/test_apigateway.py::TestWebSockets::test_websocket_api[AWS_PROXY-False-True]`
- [ ] `FAILED tests/integration/test_apigateway.py::TestWebSockets::test_websocket_api[AWS_PROXY-False-False]`
- [ ] `FAILED tests/integration/test_apigateway.py::TestHttpApis::test_lambda_request_authorizer`
- [ ] `FAILED tests/integration/test_apigateway.py::TestHttpApis::test_iam_authorizer`
- [ ] `FAILED tests/integration/test_apigateway.py::TestHttpApis::test_cognito_authorizer`
- [ ] `FAILED tests/integration/test_apigateway.py::TestHttpApis::test_lambda_request_authorizer_different_paths[gateway_request_url]`
- [ ] `FAILED tests/integration/test_apigateway.py::TestHttpApis::test_lambda_request_authorizer_different_paths[_get_request_url_custom_dns]`
- [ ] `FAILED tests/integration/test_apigateway.py::TestHttpApis::test_lambda_return_gzip_response`
- [ ] `FAILED tests/integration/test_appsync.py::TestAppSync::test_api_lambda_integration`
- [ ] `FAILED tests/integration/test_athena.py::TestAthena::test_simple_query - asse...`
- [ ] `FAILED tests/integration/test_cloudformation.py::TestCloudFormation::test_create_stack`
- [ ] `FAILED tests/integration/test_cognito.py::TestCognito::test_api_gateway_cognito_authorizer`
- [ ] `FAILED tests/integration/test_cognito.py::TestCognito::test_cognito_triggers`
- [ ] `FAILED tests/integration/test_cognito.py::TestCognito::test_user_migration_lambda`
- [ ] `FAILED tests/integration/test_docdb.py::TestDocDB::test_create_query_db - Exc...`
- [ ] `FAILED tests/integration/test_ecr.py::TestECR::test_create_delete_image - bot...`
- [ ] `FAILED tests/integration/test_eks.py::TestEKS::test_create_app_in_k3s_cluster`
- [ ] `FAILED tests/integration/test_glue.py::TestGlue::test_import_data_catalog - A...`
- [ ] `FAILED tests/integration/test_kafka.py::TestKafka::test_kafka_lambda_event_source_mapping`
- [ ] `FAILED tests/integration/test_kinesis_analytics.py::TestKinesisAnalytics::test_run_query`
- [ ] `FAILED tests/integration/test_kinesis_analytics.py::TestKinesisAnalytics::test_run_siddhi_query`
- [ ] `FAILED tests/integration/test_lambda.py::test_function_using_layer - botocore...`
- [ ] `FAILED tests/integration/test_lambda.py::test_function_using_layer_overriding_runtime`
- [ ] `FAILED tests/integration/test_lambda.py::test_custom_runtime_with_runtime_api`
- [ ] `FAILED tests/integration/test_rds.py::TestRDS::test_lambda_extensions - Asser...`
- [ ] `FAILED tests/integration/test_serverless.py::TestServerless::test_websocket_deployed`
- [ ] `FAILED tests/integration/test_xray.py::TestXRay::test_collect_lambda_traces`

**Comment by alexrashed** [comment]
### Ampere Altra (with qemu usermode emulation):
- [ ] FAILED tests/integration/test_athena.py::TestAthena::test_simple_query - asse...
- [ ] FAILED tests/integration/test_docdb.py::TestDocDB::test_create_query_db - Exc...
- [ ] FAILED tests/integration/test_eks.py::TestEKS::test_create_app_in_k3s_cluster
- [ ] FAILED tests/integration/test_glue.py::TestGlue::test_import_data_catalog - A...
- [ ] FAILED tests/integration/test_kinesis_analytics.py::TestKinesisAnalytics::test_run_query
- [ ] FAILED tests/integration/test_kinesis_analytics.py::TestKinesisAnalytics::test_run_siddhi_query

**Comment by mLupine** [comment]
Hi, as per the following:
> Lambdas running in Docker executor will likely fail. We currently do not have arm64 native Lambda runtime container images

You can use my fork of the lambci/docker-lambda project that has full support for ARM runtimes: https://github.com/mLupine/docker-lambda. Although it's heavily rewritten under the hood, it should be a plug-and-play replacement for the currently used images.

I do, however, have one more issue with arm support in 0.13.0. The Pro version of LocalStack supports fetching Lambda layers from AWS. Your lambda layer proxy probably uses the x86_64 architecture to extract the contents of a specified layer. Layers can, however, specify architectures they're compatible with and if a layer is only compatible with arm64, the proxy returns error 500. 
```
Exception: Error code 500 when requesting layer content from proxy service
```

You can reproduce the issue by trying to fetch the arm64 version of my [Git layer](https://github.com/mLupine/lambda-git). `arn:aws:lambda:eu-west-1:110966697890:layer:lambda-git-x86_64:1` works just fine, but `arn:aws:lambda:eu-west-1:110966697890:layer:lambda-git-arm64:1` fails with the aforementioned error. 

I'm not sure if I should  create a new issue, or if the report here is enough 🙈.

**Comment by alexrashed** [comment]
Actually @mLupine, for the sake of the overview we would like to use this ticket to track the failing tests and just link other issues in here. Otherwise this issue could get sidetracked quite easily.
Would it be possible for you to create a new issue and link this one in it?

**Comment by mLupine** [comment]
@alexrashed sure, just created [an issue](https://github.com/localstack/localstack/issues/4932). The first part of my message is still valid, though 😉

**Comment by steveocr** [comment]
> Most issues related to arm64 are mitigated by running additional amd64 emulation.
> This will be the case for most Apple M1 users.

How do I run it with "additional amd64 emulation"?

**Comment by alexrashed** [comment]
With "additional amd64 emulation",  @thrau was referring to [Apple Rosetta 2](https://en.wikipedia.org/wiki/Rosetta_%28software%29#Rosetta_2).
If you are running Mac on a machine with Apple silicon, you should be asked to install Rosetta 2 [automatically as soon as x84_64 code is executed](https://support.apple.com/en-us/HT211861).

If you are running a GNU/Linux system, you can enable the `bin_fmt` emulation for docker by following our guide in the docs:
https://docs.localstack.cloud/localstack/arm64-support/#using-amd64-lambda-functions-and-binaries-on-arm64

**Comment by snarky-puppy** [comment]
> UnknownEndpoint: Inaccessible host: `localhost' at port `4566'. This service may not be available in the `us-west-2' region.: connect ECONNREFUSED ::1:4566

I solved this by removing `127.0.0.1` from the `ports` section of docker-compose.yaml
```yaml
    ports:
      - "4566:4566"            # LocalStack Gateway
      - "4510-4559:4510-4559"  # external services port range
```

**Comment by joe4dev** [comment]
> It also seem there are networking issues and I get the below when trying to do a cdklocal bootstrap

@Simonl9l This is an IPv6 issue on macOS using Nodejs >=v17 (see [explanation here](https://github.com/localstack/aws-cdk-local/issues/76#issuecomment-1412590519)) and has been fixed in cdklocal [2.17.0](https://www.npmjs.com/package/aws-cdk-local/v/2.17.0) (see Issue https://github.com/localstack/aws-cdk-local/issues/76).

**Comment by joe4dev** [comment]
@mLupine @steveocr @Simonl9l 

Using our new lambda provider implementation, LocalStack now supports lambda functions running on arm64 (only on ARM hosts, see https://github.com/localstack/localstack/issues/7637), even with public ARM layers (pro feature, see https://github.com/localstack/localstack/issues/4932) 🚀 

1. Pull the latest LocalStack image `docker pull localstack/localstack-pro`
2. Start LocalStack with the environment variable [configuration](https://docs.localstack.cloud/references/configuration/) `PROVIDER_OVERRIDE_LAMBDA=v2`

**Comment by CyrilDevOps** [comment]
How can I use kinesis in localstack with my mac m2 ?
I tried 3.0.2 arm64, or 3.0.2 amd64, same problem, it is stuck doing npm install kinesis-local@0.4.4
Any fix ?

**Comment by alexrashed** [comment]
Hi @CyrilDevOps! This doesn't really seem to be related to this (tracking-)issue. Could you please create a new issue with a description on how we can reproduce your issue, and the logs you are seeing? Thanks!

**Comment by CyrilDevOps** [comment]
> Hi @CyrilDevOps! This doesn't really seem to be related to this (tracking-)issue. Could you please create a new issue with a description on how we can reproduce your issue, and the logs you are seeing? Thanks!

Made a ticket : https://github.com/localstack/localstack/issues/9969
But the issue was a limitation of our internal network that wasn't allow to connect to npm registry because we overload them, we had to go through a internal proxy, had to do some npm config in localstack container before starting kinesis.

**Comment by MSciesiek** [comment]
Hi, I encountered an issue in S3 multi-part download when using `linux/arm64` image on a M1 apple silicon. It was giving me a checksum validation error. When using `linux/amd64` the issue is not present.
Do you want to track it here or should I open a new ticket? I haven't found any other related ticket.

**Comment by alexrashed** [comment]
@MSciesiek Please create a separate issue for this one. Also please make sure to also explicitly pull the `latest` image, the issue you are describing sounds like https://github.com/localstack/localstack/issues/12187 (which should already be fixed). ;)

**Comment by MSciesiek** [comment]
@alexrashed Using the latest amd64 image indeed solved the problem. It seems that the issue was indeed fixed in #12187 .

**Comment by HarshCasper** [comment]
Due to the forthcoming changes in our distribution and development model ([read more](https://blog.localstack.cloud/the-road-ahead-for-localstack/)), this issue tracker is no longer being maintained, and this issue is now being closed. If you'd like us to take a look or need further support, please reach out through one of our support channels:

-   💬 [GitHub Discussions](https://github.com/orgs/localstack/discussions)
-   📖 [Help & Support](https://docs.localstack.cloud/aws/getting-started/help-support/)
